Add tests to make sure our test framework tolerates maven reruns even with test profiles

holly-cummins opened the pull request:
Reproducer for #46048

As expected, this passes on `main`. To my surprise, it also passes on the latest version of #34681. I was assuming I had the test wrong or wasn't testing the correct scenario (after all, never trust a test you haven't seen fail). So I checked against last week's version of #34681, and these tests failed. 

It appears a miracle has happened, and #46048 has been fixed sometime in the last week, without needing any maven changes. When I looked at this problem last time, the tests were not reloaded by the FacadeClassLoader on the second test run, so they'd run in the old classloader against a closed Quarkus application, and that worked about as well as you'd expect. This time, I've confirmed that on the second run the tests do get freshly loaded, and so everything works. 

I can also see in CI that failing LGTM tests are being rerun, and they're failing in the LGTM code each time, not in the test framework code. So that's a useful extra confirmation.

I'm not 100% sure what fixed it, but I think it's perhaps a combination of @aloubyansky's  suggestions about cleaner test file discovery, and my work to close classloaders and leak less.

holly-cummins committed:
Add tests to make sure our test framework tolerates -Dsurefire.rerunFailingTestsCount

aloubyansky commented on integration-tests/maven/src/test/java/io/quarkus/maven/it/SurefireRerunFailedTestsIT.java:
```diff
@@ -0,0 +1,64 @@
+        final RunningInvoker running = new RunningInvoker(testDir, false);
+
+        // trigger mvn verify -Dsurefire.rerunFailingTestsCount=2
+        final String[] mvnArgs = new String[] { "verify", "-Dsurefire.rerunFailingTestsCount=3" };
```
Should the comment be aligned with the actual argument value passed?

holly-cummins replied:
I'll update the comment to something generic so that we don't need to double-maintain if we make changes in the test.

holly-cummins commented:
Rebasing in the hopes that the problem with the build reports has been fixed upstream.

gsmet commented:
There's something odd here, the build report is failing with the flake message being null, which is the very very first time I actually see that.

holly-cummins commented:
> There's something odd here, the build report is failing with the flake message being null, which is the very very first time I actually see that.

Ohh, that's got to be related to my changes, then. I saw Qute and assumed it wasn't me. Although I don't know why it would be happening, because the failure is just an ordinary test failure. I'll have a look to see if there's some exception message I've failed to fill in somewhere. 

I guess it also means my test will turn up on every flake report, which is a bit irritating, even if I make sure the message is "deliberate failure" or something informative. I wonder if there's a property I can set to exclude it?

holly-cummins commented:
I think the problem might have been caused by failing with thrown `InvocationTargetExceptions`, which (maybe?) have a null message. I've adjusted the fake-flaky test to unwrap the exception, and squashed.

gsmet commented:
I deployed a fix to both the Bot and the Build Reporter Action.

holly-cummins commented:
Either your fix or my unwrapping of the exception sorted it out, @gsmet. Do you think we should try and exclude those failures from the build report? Otherwise that's going to be annoyingly noisy.

gsmet commented:
Ah, my, yes, this is going to be very annoying.

Let's not merge right away, then. I'll mark it as draft for now.

holly-cummins commented:
We probably want some kind of "expected flakes" config on the build reporter.

gsmet commented:
Yes. We need to add some infrastructure for that. I'll have a look soon.

holly-cummins commented:
@gsmet, I don't want to lose track of this. Did I understand right that you were having a look to understand why your fix didn't have the expected effect? Or is there an action with me?

holly-cummins commented:
In case anyone else is wondering why these tests are green even though main CI is having problems with flaky tests reruns (https://github.com/quarkusio/quarkus/issues/51269). It turns out the current problem in main was caused by https://github.com/quarkusio/quarkus/pull/50092, which bumped surefire from 3.5.3 to 3.5.4. These tests don't lock the version of surefire to the one used by the rest of the build. In hindsight, that's a mistake, but it's one I can't fix until I also fix #51269, or the tests would start failing.
